mlir/lib/Quantum/Transforms/ChainedSelfInversePatterns.cpp

```diff
@@ -30,24 +29,33 @@ namespace {
 struct ChainedHadamardOpRewritePattern : public mlir::OpRewritePattern<CustomOp> {
     using mlir::OpRewritePattern<CustomOp>::OpRewritePattern;
 
-    /// We check if the operation and it's parent are hadamard operations. If so, replace op
-    /// with it's "grandparent".
+    /// We simplify consecutive Hermitian quantum gates and simplifies them.
+    /// Hermitian gates are self-inverse and applying the same gate twice in succession
+    /// cancels out the effect. This pattern rewrites such redundant operations by
+    /// replacing the operation with its "grandparent" operation in the quantum circuit.
     mlir::LogicalResult matchAndRewrite(CustomOp op, mlir::PatternRewriter &rewriter) const override
     {
-        LLVM_DEBUG(dbgs() << "Simplifying the following hadamard operation:\n" << op << "\n");
-        if (op.getGateName().str() != "Hadamard")
-            return failure();
+        LLVM_DEBUG(dbgs() << "Simplifying the following operation:\n" << op << "\n");
 
-        ValueRange qbs = op.getInQubits();
-        auto parentHadamard = dyn_cast<CustomOp>(qbs[0].getDefiningOp());
+        mlir::StringSet<> HermitianOps = {"Hadamard", "PauliX", "PauliY", "PauliZ", "CNOT",
```

Comment:
Since this set is not supposed to be changed, I think it would be better to make it a `const` member of the struct, instead of making a new one every time the `matchAndRewrite` is called.

In the actual pass the pattern class is only created once, and this set really just needs to be created once as well: https://github.com/PennyLaneAI/catalyst/blob/main/mlir/lib/Quantum/Transforms/remove_chained_self_inverse.cpp#L74

Thoughts?

Reply:
Good suggestion! I actually will make it a static const outside the struct since it is not really related to the struct as well. This way, it will definitely get created only once.